### PR TITLE
Thorough testing of Firestore DB

### LINKED ISSUE #246: Finish testing FirestoreDatabase
The goal is to test all FirestoreDatabase. So far, only the first functions are tested, not the one involving transactions.

### PR DESCRIPTION
**Pull request by nathanduchesne**
If merged, this PR will add tests to all our calls to Firestore DB, showing a great capacity of mocking.
This should drastically increase our repo coverage.

### CODE REVIEW

**Comment by nathanduchesne** [comment]
Initially, this PR aimed to test Firestore database to increase our coverage. It would have been nice to mock all FirestoreDB calls for the following reasons:
1. Our test suite could have been run from offline devices
2. The number of actual calls to the Google firestore database would remain as low as possible.

However, after trying to mock transactions for the past week, I realized this would much more work than what was initially planned for. These tests therefore actually call the database, but have the advantage of actually testing the behaviour, whereas mocking would have just allowed us to test input and output, which in transactions, is often just null. Moreover, some pre-existing tests that call the database already have been merged to master, so rejecting these would be harsh.
To answer 1), the CI runs all tests online anyways, and since this test suite concerns a database, why test it on the emulator anyways?

Please note this PR drastically increases the total project coverage, which is the initial goal.

File: app/src/androidTest/java/com/github/fribourgsdp/radio/activities/FirestoreDatabaseTest.kt [hunk 1]
```diff
@@ -113,14 +132,430 @@ class FirestoreDatabaseTest {
...
+
+    @Test(expected = Exception::class)
+    fun getNonExistingGameSettingsFails() {
+        val fireDb = FirestoreDatabase()
+        Tasks.await(fireDb.getGameSettingsFromLobby(3333))
+    }
+
+    @Test(expected = Exception::class)
+    fun addingUserToNonExistingLobbyFails() {
+        val fireDb = FirestoreDatabase()
```
Comment (BartLucien) [inline app/src/androidTest/java/com/github/fribourgsdp/radio/activities/FirestoreDatabaseTest.kt:158]: It seems that you are always calling these line in all the test

**Review by BartLucien — COMMENTED** [review]
Overall, the testing looks fine and it greatly increases the overall coverage. Whether it is better to have tests that rely on external services or no test at all, I cannot say. In our case, Google seems like a stable company to rely on. However, if there are tests that rely on API which could be mocked, like the one from previous weeks, these should be avoided and corrected in the future.

File: app/src/androidTest/java/com/github/fribourgsdp/radio/activities/FirestoreDatabaseTest.kt [hunk 2]
```diff
@@ -113,14 +132,430 @@ class FirestoreDatabaseTest {
...
+    @Test(expected = Exception::class)
+    fun resetGameMetadataFailsIfWrongId() {
+        val fireDb = FirestoreDatabase()
+        Tasks.await(fireDb.resetGameMetadata(987987, "123"))
+    }
+
+    @Test
+    fun getLobbyIdWorks() {
+        `when`(mockSnapshot.exists()).thenReturn(true)
+        `when`(mockSnapshot.get("current_id")).thenReturn(3L)
```
[the remark this answers is not in the document]
Comment (nathanduchesne) [reply app/src/androidTest/java/com/github/fribourgsdp/radio/activities/FirestoreDatabaseTest.kt:215]: I've done it in the previous commit!

File: app/src/androidTest/java/com/github/fribourgsdp/radio/activities/FirestoreDatabaseTest.kt [hunk 3]
```diff
@@ -113,14 +140,393 @@ class FirestoreDatabaseTest {
...
+    }
+
+    @Test(expected = Exception::class)
+    fun disablingLobbyFailsIfWrongId() {
+        Tasks.await(fireDb.disableLobby(unusedId))
+    }
+
+    @Test(expected = Exception::class)
+    fun makeSingerDoneFailsIfWrongId() {
+        Tasks.await(fireDb.makeSingerDone(unusedId, "123"))
```
[the remark this answers is not in the document]
Comment (nathanduchesne) [reply app/src/androidTest/java/com/github/fribourgsdp/radio/activities/FirestoreDatabaseTest.kt:194]: I'll fix this asap.

File: app/src/androidTest/java/com/github/fribourgsdp/radio/activities/FirestoreDatabaseTest.kt [hunk 4]
```diff
@@ -113,14 +140,393 @@ class FirestoreDatabaseTest {
...
+    }
+
+    @Test(expected = Exception::class)
+    fun registerEmptyPlaylistNameThrowsExcep() {
+        Tasks.await(db.registerPlaylist(Playlist("")))
+    }
+
+    @Test
+    fun openGameWorks() {
+        db.openGame(3L)
```
[the remark this answers is not in the document]
Comment (nathanduchesne) [reply app/src/androidTest/java/com/github/fribourgsdp/radio/activities/FirestoreDatabaseTest.kt:290]: You're write I'll change this.

**Review by Flamantrose321 — COMMENTED** [review]
In general it's fine, just can you also test the FirestoreDatabase line 290?

**Comment by nathanduchesne** [comment]
> In general it's fine, just can you also test the FirestoreDatabase line 290?

I assume you mean better test line 290 in the tests (the fact I only called db.openGame), I have fixed this in the latest commit and go through a deeper check. Hope this is enough.

**Comment by Flamantrose321** [comment]
> > In general it's fine, just can you also test the FirestoreDatabase line 290?
> 
> I assume you mean better test line 290 in the tests (the fact I only called db.openGame), I have fixed this in the latest commit and go through a deeper check. Hope this is enough.

Yes, I did make a typo.
Okay, perfect.

**Review by Flamantrose321 — APPROVED** [review]
He answered my question. It's good for me.

**Review by BartLucien — APPROVED** [review]
I have nothing to say except for the comment I already made. This can be merged for me.

**Pull request merged**